### ISSUE TITLE
Crash on exporting with tcl plugin

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#1019]
# Describe what you wanted to do

```
$ cd src/plugins/xmltool/xmltool
$ kdb import user/tests/dump xmltool < dump.xml
$ kdb export user/tests/dump tcl
{
        {
*** Error in `kdb': munmap_chunk(): invalid pointer: 0x000000000119c490 ***
zsh: abort (core dumped)  noglob kdb export user/tests/dump tcl
```

Exporting with ni plugin (or xmltool or dump) works without issues.
Seems like tcl plugin creates broken KeySets refering to deleted or non-existing keys?
# System Information
- Elektra Version: master
# Backtrace

```
#0  keyDecRef (key=key@entry=0x40) at /home/markus/Projekte/Elektra/current/src/libs/elektra/key.c:583
#1  0x00007f65b0ca7e58 in ksClose (ks=ks@entry=0x1068380) at /home/markus/Projekte/Elektra/current/src/libs/elektra/keyset.c:2393
#2  0x00007f65b0ca7eb3 in ksDel (ks=0x1068380) at /home/markus/Projekte/Elektra/current/src/libs/elektra/keyset.c:356
#3  0x0000000000445af7 in ~KeySet (this=0x7ffd9efb2300, __in_chrg=<optimized out>) at /home/markus/Projekte/Elektra/current/src/bindings/cpp/include/keyset.hpp:568
#4  ExportCommand::execute (this=<optimized out>, cl=...) at /home/markus/Projekte/Elektra/current/src/tools/kdb/export.cpp:39
#5  0x0000000000417427 in main (argc=4, argv=0x7ffd9efb26a8) at /home/markus/Projekte/Elektra/current/src/tools/kdb/main.cpp:14
```
# Valgrind

``````
{
        {
==23667== Invalid write of size 8
==23667==    at 0x4E3A835: ksRewind (keyset.c:1219)
==23667==    by 0x4E3BE4D: ksClose (keyset.c:2390)
==23667==    by 0x4E3BEB2: ksDel (keyset.c:356)
==23667==    by 0x445AF6: ~KeySet (keyset.hpp:568)
==23667==    by 0x445AF6: ExportCommand::execute(Cmdline const&) (export.cpp:39)
==23667==    by 0x417426: main (main.cpp:145)
==23667==  Address 0xce41688 is 24 bytes inside a block of size 48 free'd
==23667==    at 0x4C29E90: free (vg_replace_malloc.c:473)
==23667==    by 0x4E3BEBC: ksDel (keyset.c:357)
==23667==    by 0xD3539D2: ???
==23667==    by 0x5277209: kdb::tools::Plugin::set(kdb::KeySet&, kdb::Key&) (plugin.cpp:399)
==23667==    by 0x4456FB: ExportCommand::execute(Cmdline const&) (export.cpp:63)
==23667==    by 0x417426: main (main.cpp:145)
==23667==
[...]

```==23667== Invalid free() / delete / delete[] / realloc()
==23667==    at 0x4C29E90: free (vg_replace_malloc.c:473)
==23667==    by 0x4E3BEBC: ksDel (keyset.c:357)
==23667==    by 0x445AF6: ~KeySet (keyset.hpp:568)
==23667==    by 0x445AF6: ExportCommand::execute(Cmdline const&) (export.cpp:39)
==23667==    by 0x417426: main (main.cpp:145)
==23667==  Address 0xce41670 is 0 bytes inside a block of size 48 free'd
==23667==    at 0x4C29E90: free (vg_replace_malloc.c:473)
==23667==    by 0x4E3BEBC: ksDel (keyset.c:357)
==23667==    by 0xD3539D2: ???
==23667==    by 0x5277209: kdb::tools::Plugin::set(kdb::KeySet&, kdb::Key&) (plugin.cpp:399)
==23667==    by 0x4456FB: ExportCommand::execute(Cmdline const&) (export.cpp:63)
==23667==    by 0x417426: main (main.cpp:145)
==23667== 
==23667== Invalid read of size 8
==23667==    at 0x4175F8: main (main.cpp:174)
==23667==  Address 0xd563800 is not stack'd, malloc'd or (recently) free'd
==23667== 
==23667== 
==23667== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==23667==  Access not within mapped region at address 0xD563800
==23667==    at 0x4175F8: main (main.cpp:174)
==23667==  If you believe this happened as a result of a stack
==23667==  overflow in your program's main thread (unlikely but
==23667==  possible), you can try to increase the size of the
==23667==  main thread stack using the --main-stacksize= flag.
==23667==  The main thread stack size used in this run was 8388608
``````

**Comment by markus2330** [comment]
Any solution in sight? Can you fix this for 0.8.19?

**Comment by sanssecours** [comment]
> Any solution in sight? Can you fix this for 0.8.19?

Not yet, sorry. I will take a look at this after I finish my exposé, if that is okay with you.

**Comment by markus2330** [comment]
Yes, is okay!

**Comment by markus2330** [comment]
Isn't this implicitly solved via your new plugin? (So we only need tcl to depend on base64?)

**Comment by sanssecours** [comment]
> Isn't this implicitly solved via your new plugin?

I do not think the new “meta mode” feature of the Base64 plugin solves the problem. However the “old” escaping mode of the plugin should come in handy. If we require `null` and `base64`(or `binary`) in the `tcl` plugin reading and writing should work without any problems. For example, the following works:

```sh
kdb mount test.tcl user/tests tcl null binary
kdb import user/tests/dump xmltool < src/plugins/xmltool/xmltool/dump.xml
kdb ls user/tests/dump
#> user/tests/dump/.HiddenBinaryKey
#> user/tests/dump/.HiddenDirectoryKey
#> user/tests/dump/.HiddenStringKey
#> user/tests/dump/PerfectBinaryKey
#> user/tests/dump/PerfectDirectoryKey
#> user/tests/dump/PerfectStringKey
#> user/tests/dump/Ug.ly:Bin@aryKey
#> user/tests/dump/Ug.ly:Dir@ectoryKey
#> user/tests/dump/Ug.ly:St@rin.gKey
kdb get user/tests/dump/PerfectBinaryKey
#> \x42\x69\x6e\x61\x72\x79\x56\x61\x6c\x75\x65\x0
kdb get user/tests/dump/Ug.ly:Bin@aryKey
#>
```

. Anyway since `kdb export` only [supports a single plugin](https://github.com/ElektraInitiative/libelektra/pull/1194#issuecomment-269909675):

```sh
kdb export user/tests tcl
#> {
#>	{
#>		user/tests/dump/.HiddenBinaryKey =
Sorry, I crashed by the signal SIGSEGV
This should not have happened!

Please report the issue at https://issues.libelektra.org/
fish: 'kdb export user/tests tcl' terminated by signal SIGABRT (Abort)
```

we are still out of luck, even if we add `null` and `binary` to `info/needs`.

**Comment by markus2330** [comment]
>  If we require null and base64(or binary) in the tcl plugin reading and writing should work without any problems.

Perfect, requiring null and binary sounds exactly like it should be. It might be needed for other plugins, too. (E.g. simpleini?)

> . Anyway since `kdb export` only [supports a single plugin](https://github.com/ElektraInitiative/libelektra/pull/1194#issuecomment-269909675):

Good point, we should open a new issue for this to make clear that TCL (and other plugins in need of null+binary) can crash on exports because #1229 is not yet implemented. (I do not see #1229 in the scope of 0.8.20).
